### ISSUE TITLE
Protobuf communications

### ISSUE THREAD

**Issue by timothysc** [kubernetes/kubernetes#8132]
re: #3338 

Post 1.0 and etcd v3. 
It would be worth while to investigate the tradeoffs for using protobuf for internal communication whose external representation is json for performance and scale out reasons.

**Comment by bgrant0607** [comment]
There's wire format (json, proto, [bson](http://bsonspec.org/), ...), IDL (tagged Go, proto, ...), server framework (go-restful, grpc, ...), and client generation (hand-written, Swagger codegen, grpc, ...).

Some issues with proto/grpc to consider:
* Select schema based on apiVersion and kind (and prefix, etc., if we add those)
* Mapping to CRUD/OO API
* Auto-generated support for JSON/REST (and YAML, via conversion to JSON) (e.g., https://github.com/gengo/grpc-gateway)
* Areas of incompatibility with current API schema
* Older clients with newer apiserver (#3955)
* How to support instances of custom parsing (time, resources, IntOrString, ...)
* Would want service per object, effectively breaking up the API (#1744)
* Dynamic API Plugins (#991, #703, #3201)
* Swagger support
* Tagging for validation (#8116)
* Auto-generation of common methods (e.g., watch)
* Patch and config reconciliation (#1702)
* Interaction with defaulting, conversion, and validation
* Explicit default values on the wire
* Distinguishing unset vs. language defaults
* Strict validation
* Descriptive error messages for unknown fields
* HTTP functionality (redirects, proxy, retries, status codes, etc.)
* Use from Javascript (for GUI)

**Comment by erictune** [comment]
We are using pointer-to-scalar to distinguish unset from zero value in several places.  Proto does not support this.

**Comment by davidopp** [comment]
Google's proto3 supports some kind of inter-operation with JSON that I did not take the time to read about in detail: https://developers.google.com/protocol-buffers/docs/proto3#json

**Comment by smarterclayton** [comment]
I looked into it - I think our use of pointers to denote optional an unspecified fields may cause an issue

**Comment by smarterclayton** [comment]
may *still* cause an issue even with proto3 (noting Eric already pointed out the issue)

**Comment by thockin** [comment]
I don't know if we have any leverage to fix this issue - it sounds very
intentional that they eliminated both default values AND
presence-detection.  Maybe we could do something clever with JSON-to-proto
encoding, but that will make the proto-native interface ugly.

LOSE

**Comment by pherl** [comment]
Proto3 actually supports field presence for primitive fields in a similar manner. i.e. https://github.com/google/protobuf/blob/master/src/google/protobuf/wrappers.proto 

During the JSON conversion, unset value will not be serialized (as the proto3 spec), for the wrapper type, if you set the actual value to 0, the JSON serializer will print a '0' in the output.

There is also a Value proto defined in https://github.com/google/protobuf/blob/master/src/google/protobuf/struct.proto with which, you can distinguish empty array from null array.

**Comment by bgrant0607** [comment]
Thanks, @pherl. So we'd use wrapper messages when we wanted to distinguish unset from type default.

1 issue resolved, ~20 to go.

**Comment by thockin** [comment]
So someone who was writing protobuf directly will have to say
spec.replicas.value = 0 instead of spec.replicas ?  It's a passable
workaround, but it's pretty ugly.

**Comment by bgrant0607** [comment]
There are many other issues that would make proto ugly, anyway. For instance, I'd guess that we'd need a generic object with metadata, which all specific objects would then extend (per api group, per api version). Fields with custom parsing would need to be represented as strings. And so on. With proto, implementation concerns bleed into the API heavily than with json.

I'm skeptical that we could use proto as our IDL (interface definition language). We'd probably need to generate it.

**Comment by smarterclayton** [comment]
That's good to know - did not see those while looking.  Is there a cost associated with Any?

**Comment by timothysc** [comment]
It seems that @bgrant0607 is pretty set against.  The primary driver of the ticket was performance related issues that we had seen, namely in decoding, parsing, GC, etc.  

FWIW when we recently tested load balancing api-servers, and a lot of the risk is mitigated but performance is still sub-optimal, especially as we approach 10^3+ in the future.  I'm ok with that, so long folks agree on the prescription.   

@wojtek-t, @filbranden  - thoughts?

**Comment by bgrant0607** [comment]
@timothysc Switching to proto and/or grpc would be a pretty drastic change, likely requiring a rewrite of all our API infrastructure, as well as breaking API compatibility. We should continue to pursue lower-hanging fruit in the short term, and consider proto more seriously when we're planning to overhaul the API infrastructure for other reasons.

**Comment by smarterclayton** [comment]
At least at the serialization level we should be able to isolate a protobuf to only our external serialization struct and support it as an alternative transport.  It does come down as you noted to being able to, from an existing versioned struct, being able to generate a 100% accurate protobuf representation of the versioned struct and recover it.  That's something that is a good candidate for experimentation with something like gogoprotobuf because we already have the infrastructure to test round trip fidelity from serialization and back.

But it's definitely a chunk of work, and would have to represent a significant wire serialization speed up.

----- Original Message -----

**Comment by wojtek-t** [comment]
cc @fgrzadkowski 

I think that one thing that we should do in this area is to just change the communication between apiserver and etcd to protocol-buffer-based. If we just do this, we don't need to even translate our api objects to protobuffers (and still store serialized json objects in etcd). However, my feeling is that the gain from just switching etcd<->apiserver communication to protobuffers would itself be a huge win.

Also, I agree with @bgrant0607 that we should start with some other things that could have significant performance impact first before just doing it. So as far as my opinion is concerned I would start with:

1. fixing some issues we are aware of (e.g. implementing watch at the apiserver level, so that we don't sent/deserialize the same data multiple times - each Kubelet is watching for pods assigned to it, but we listen to all pods for every single Kubelet in apiserver and do the filtering at the apiserver level). We can probably think of other problems like that that we can solve. 
BTW, I'm planning to start working on it once I'm done with all the work I need to do for 1.0] - @lavalamp 

2. Change to communication between apiserver and etcd to protobuf-based once v3 etcd api exists.

3. Only if that's not enough, start thinking about making supporting protobuf-based api in general (which as you see from the discussion above will definitely not be straightforward).

**Comment by bgrant0607** [comment]
We should also consider eliminating common-case conversions, as I just proposed in #8190.

**Comment by wojtek-t** [comment]
@bgrant0607 - this is at least worth considering. But in my above post, it will be the part of (1).

**Comment by timothysc** [comment]
At this point, does it make sense to create a laundry list ticket for post 1.0 performance ideas, and reference this ticket as one option. 

We are starting to create a laundry list in this ticket...

**Comment by bgrant0607** [comment]
No, let's put the laundry list elsewhere. For API- and client-related things, that's #8190.

**Comment by lavalamp** [comment]
One thing I would like to try is to make an alternative BSON (http://bsonspec.org/) encode/decode mechanism. With any luck, this would get the benefits of a binary format while being extremely easy to fit into our current system, since it's literally a binary form of JSON.

There's a decent go library for it: http://godoc.org/labix.org/v2/mgo/bson

But I expect we'd need to go the generated code route.

**Comment by davidopp** [comment]
+1 for having a single place to record all of the post-1.0 performance ideas (but not using this issue to do it). The list can point into #8190, this issue, etc.

I can set something up in docs/proposals unless someone has a better idea.

**Comment by wojtek-t** [comment]
+1 for having a single place for post-1.0 plans/ideas

**Comment by davidopp** [comment]
I was thinking smaller-scale -- just for brainstorming post-1.0 performance/scalability plans. *All* post-1.0 plans is a pretty tall order. :)

**Comment by wojtek-t** [comment]
@davidopp - sorry for not being clear. I didn't mean "All" post-1.0 plans - I meant "all" that so far came to our mind. This should be extensible list.

**Comment by wojtek-t** [comment]
Basically - to have all discussions in a single place.

**Comment by bgrant0607** [comment]
Filed #10224 to end the chatter (6+ comments!) in this issue about filing another issue.

**Comment by davidopp** [comment]
Facebook blogs about performance improvement switching from JSON to something called FlatBuffer, which is similar to Proto Buffers (and also an open-source project from Google). This is the context of mobile so may not be directly applicable but seems to make some reasonable points.

https://code.facebook.com/posts/872547912839369/improving-facebook-s-performance-on-android-with-flatbuffers/

**Comment by smarterclayton** [comment]
I looked at flat buffer the other day - I think we could do the schema
generation directly from our external structs, but don't have the time to
prototype it.  I don't see any reason we couldn't serialize external to
flat buffer based on a cursory investigation.

On Aug 2, 2015, at 5:56 PM, David Oppenheimer <notifications@github.com>
wrote:

Facebook blogs about performance improvement switching from JSON to
something called FlatBuffer, which is similar to Proto Buffers (and also an
open-source project from Google). This is the context of mobile so may not
be directly applicable but seems to make some reasonable points.

https://code.facebook.com/posts/872547912839369/improving-facebook-s-performance-on-android-with-flatbuffers/

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/issues/8132#issuecomment-127075097>
.

**Comment by mikedanese** [comment]
What is the cost/benifit of using flatbuffers vs protobuf?

**Comment by smarterclayton** [comment]
Flatbufs in theory can do zero allocation during deserialize, and is
slightly faster in some benchmarks.

On Aug 2, 2015, at 11:18 PM, Mike Danese <notifications@github.com> wrote:

What is the cost/benifit of using flatbuffers vs protobuf?

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/issues/8132#issuecomment-127110162>
.

**Comment by bgrant0607** [comment]
Update from a discussion yesterday:
* We don't have to use the grpc client/server framework in order to use proto, and we don't need to use proto with grpc. They are independent choices. This issue should be just about the encoding, not about the API framework.
* In proto3, we'd need to use messages boxing primtive types for optional fields rather than directly using primitive types. 
* Proto2 is still an option. It supports distinguishing unset from default. 
* We'd probably want to autogenerate the proto schema from our Go structs, rather than use protobuf as our IDL. Alternatively, we could treat proto representations as distinct API versions and manually maintain the proto definitions.
* Either way, we'd need to write a new conversion generator, to convert from the proto-generated Go structs to the internal rep and vice versa. 
* It was suggested that we use "dynamic types" in order to decode based on apiVersion and kind. Proto3 supports them in some languages. Need to check status for Go. This might mean using non-generated decoders, so we may need to do something else for this, like have a wrapper message with just apiVersion, kind, and an any field, separate from the actual resource types.
* Messages are length-prefixed, so it is possible to skip messages without decoding them in order to defer decoding, which is presumably used for extensions/any.
* There's no standard API for fetching the proto schema directly, but there is a serialized (proto) representation of it, so we could create a swagger-like API to fetch it.
* I assume we'd continue to just use json and yaml in kubectl. Proto would be used by controllers and our other automated components -- users of our current "unversioned client".

There are ruby, python, and java implementations:
https://github.com/google/protobuf

The go implementation is in another org/repo:
https://github.com/golang/protobuf

We'd be interested in hearing from others in the community about how they feel about protobufs vs. any other binary formats.

**Comment by bgrant0607** [comment]
Also:

* We would not use proto2 support for implicit (non-serialized) defaults.
* We could not support a proto-based encoding type or API version for PATCH, at least not initially
* Clients compiled with static types would get the benefit of compile-time errors, similar to clients of our unversioned client library

**Comment by smarterclayton** [comment]
Re: generating Idl from go structs: +1.

Flat buffers has some space efficiency but the trade off seems smaller
community.  However, as a storage format once we've done one idl generator
another is easier.  Protobufs definitely is a mind share winner.



On Sep 2, 2015, at 6:10 PM, Brian Grant <notifications@github.com> wrote:

Update from a discussion yesterday:

   - We don't have to use the grpc client/server framework in order to use
   proto, and we don't need to use proto with grpc. They are independent
   choices. This issue should be just about the encoding, not about the API
   framework.
   - In proto3, we'd need to use messages boxing primtive types for
   optional fields rather than directly using primitive types.
   - Proto2 is still an option. It supports distinguishing unset from
   default.
   - We'd probably want to autogenerate the proto schema from our Go
   structs, rather than use protobuf as our IDL. Alternatively, we could treat
   proto representations as distinct API versions and manually maintain the
   proto definitions.
   - Either way, we'd need to write a new conversion generator, to convert
   from the proto-generated Go structs to the internal rep and vice versa.
   - It was suggested that we use "dynamic types" in order to decode based
   on apiVersion and kind. Proto3 supports them in some languages. Need to
   check status for Go. This might mean using non-generated decoders, so we
   may need to do something else for this, like have a wrapper message with
   just apiVersion, kind, and an any field, separate from the actual resource
   types.
   - Messages are length-prefixed, so it is possible to skip messages
   without decoding them in order to defer decoding, which is presumably used
   for extensions/any.
   - There's no standard API for fetching the proto schema directly, but
   there is a serialized (proto) representation of it, so we could create a
   swagger-like API to fetch it.
   - I assume we'd continue to just use json and yaml in kubectl. Proto
   would be used by controllers and our other automated components -- users of
   our current "unversioned client".

There are ruby, python, and java implementations:
https://github.com/google/protobuf

The go implementation is in another org/repo:
https://github.com/golang/protobuf

We'd be interested in hearing from others in the community about how they
feel about protobufs vs. any other binary formats.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/8132#issuecomment-137260026>
.

**Comment by bgrant0607** [comment]
An advantage of BSON is that it's self-describing, which is nice for dynamic types, type extensions, annotations, etc., esp. for logging and debugging. Otherwise, such data would be opaque except to the client(s) that understood it. That has a significant cost, obviously.

**Comment by smarterclayton** [comment]
One of our near term motivations is going to push to extremely high
densities of keys in etcd - millions of keys in some cases - to support
long tail density on non-primary keys (90% inactive users).  Any additional
storage efficiency we can gain (etcd3 spilling keys to disk and using
binary keys and values, protobufs or flatbuffers, compression, wire size)
helps us get there.  So we personally would be more inclined to go towards
schemaless, but I assume the infra we lay down would reduce cost for any
format.

The first one is the hardest to wire through because of the assumption of a
single codec in some code paths.


On Sep 2, 2015, at 6:53 PM, Brian Grant <notifications@github.com> wrote:

An advantage of BSON is that it's self-describing, which is nice for
dynamic types, type extensions, annotations, etc., esp. for logging and
debugging. Otherwise, such data would be opaque except to the client(s)
that understood it. That has a significant cost, obviously.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/8132#issuecomment-137266659>
.

**Comment by bgrant0607** [comment]
We have found benefits to compressing even protobufs in our storage format.

**Comment by bgrant0607** [comment]
Though that may need to be pushed into etcd itself to get the full benefit.

**Comment by bgrant0607** [comment]
At some point we'll need to automate the change of storage groups/versions, which is done by a script today. We should also think about how to change the storage format of a running cluster from json to a binary format. Unlike api versions, we don't represent the storage format explicitly.

**Comment by timothysc** [comment]
@smarterclayton At millions of keys, I often question offloading any non-operational data. 

@xiang90 ^ fyi re: ticket.

**Comment by smarterclayton** [comment]
Yes, authentication, authorization, and such can and should be decoupled on
top.  However, services and deployments and replication controllers and
endpoints and scheduled jobs and pod autoscalers and pod metrics and jobs
(builds, scheduled jobs, and run once jobs) are likely to fan out 1:1 with
number of namespaces/apps, and those are much harder to split.  The more
this becomes a platform for resource sharing (vs merely existing an
application deployment target) the larger the number of "applications",
which has a much higher fan out in # of keys than a single app with 100 or
1000 pods.

Add in idling (autoscaling to zero) and it becomes economical to build and
try out apps and have them linger indefinitely.  There are certainly
approaches that reduce the impact of the long tail of apps, but they have
high costs (sharding, partitioning, tricks).  I'd assert that optimizing
the read storage path through etcd and the internal memory is far more
valuable to invest time in in the near term, because it pays dividends in
all paths.  But I agree we don't necessarily want to over optimize for a
scenario most people don't have.

On Sep 3, 2015, at 4:20 PM, Timothy St. Clair <notifications@github.com>
wrote:

@smarterclayton <https://github.com/smarterclayton> At millions of keys, I
often question offloading any non-operational data.

@xiang90 <https://github.com/xiang90> ^ fyi re: ticket.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/8132#issuecomment-137562888>
.

**Comment by wojtek-t** [comment]
>
I'd assert that optimizing
the read storage path through etcd and the internal memory is far more
valuable to invest time in in the near term, because it pays dividends in
all paths.  But I agree we don't necessarily want to over optimize for a
scenario most people don't have.

+1

**Comment by bgrant0607** [comment]
@timothysc I didn't understand your latest comment. Are you referring to offloading resource status, events, and metrics?

**Comment by bgrant0607** [comment]
We've tried to keep the state shardable by at least kind, and perhaps namespace. Moving out auth-related resources would just be a special case of that.

**Comment by bgrant0607** [comment]
But now we're getting away from communication and into storage optimizations.

**Comment by davidopp** [comment]
> We've tried to keep the state shardable by at least kind

I think the granularity of sharding may have to be coarser than kind (i.e. in some cases 2 or more kinds would have to be in the same shard) if/when we add multi-object transactions. Multiple parallel (e.g. framework) schedulers is probably the first use case for this, assuming we don't want to pay the cost of going all the way out to the kubelet to find out a binding conflicted -- we'd need bindings to have a precondition on the version of some other object. Anyway regardless of what you think of that particular example, it's likely we're going to need multi-object transactions.

I would think namespace would be shardable (except of course for objects that aren't namespaced), unless I'm missing something.

**Comment by timothysc** [comment]
@bgrant0607 namely events, and metrics.  Which could each be their own introspective service.

**Comment by therc** [comment]
I mentioned this at Kubecon, but adopting gRPC would also bring in Census with it, one day. Most Googlers I have talked to seem unaware that the C++ gRPC codebase contains incomplete code and documentation for a Census reimplementation:

https://github.com/grpc/grpc/tree/master/src/core/census

I'm assuming the Go tree would have the same, eventually. A full Census implementation would be very helpful for performance analysis and troubleshooting.

Also, the Flatbuffer author was around for lunch, since he happens to live a few blocks away from the Kubecon venue, and he said he'd be happy to help if more work is needed to tailor the code to projects' needs.

**Comment by bgrant0607** [comment]
Yet another binary format: http://ubjson.org/

**Comment by timothysc** [comment]
xref #16649

**Comment by davidopp** [comment]
containerd uses gRPC
https://github.com/docker/containerd

**Comment by timothysc** [comment]
+ etcd3 is also making the switch.

**Comment by lavalamp** [comment]
@smarterclayton has done a lot of work on this but I don't think we're going to land it for 1.2.

**Comment by smarterclayton** [comment]
No, no plan for 1.2.  We'll be iterating after 1.3 opens.

**Comment by wojtek-t** [comment]
@smarterclayton - I'm happy to help with that once I'm done with my work for 1.2 - we can sync about it then (if you need help)

**Comment by smarterclayton** [comment]
I'll definitely want to get someone else familiar with this and point out
alternative patterns / approaches as well.  I'm certain there's things in
the proto i'm generating that we should argue on, and then some of the
patterns of how we serialize deserve a deeper discussion.

I also am looking at generating defaulters distinct from conversions (so we
can avoid converting v1 -> internal -> v1) which I will want review on in
the 1.3 frame.

**Comment by wojtek-t** [comment]
@smarterclayton ok, I will setup some VC once we are done with 1.2; thanks!

**Comment by bgrant0607** [comment]
@smarterclayton Is there a writeup of the current status on proto support somewhere?

**Comment by smarterclayton** [comment]
I need to create a proposal, which I will do in the next week

Quick summary:

* We have a fully developed tool path that has been exercised against
recent API objects that can turn Go structs into an Protobuf 2 IDL, and
then gogo-protobuf can turn that IDL into a byte for byte compatible Go
struct, then generate Marshallers / Unmarshallers that demonstrate a 10x
CPU reduction and ~5-10x memory allocation reduction against ugorji, the
generated protobufs are on average 2x smaller than JSON when used with
fuzzed objects
* No effort at IDL optimization has been made - all fields are represented
as semantically as possible (no enumerations or optimization of ordering or
types)
* hack/updated-generated-protobufs.sh works, but is currently blocked by
the "Size" field on Image info needing to become "ReportedSize" to not
conflict with the generated Protobuf field
* the generation/verify flow was dependent on kube-cross being pushed to
GCR so developer workflows would have access to a container where they
could recompile the protoc without needing to install protoc 3.0.0 on their
machine
* The majority of the code to support multiple serializers on the server,
content-negotiation, and client side request of content types is merged to
master but unused except for yaml
* A protobuf serializer is open waiting for review
* A protobuf watch implementation is ready, but has some loose ends related
to the exact watch protocol we want to expose
* End to end tests have not been run against the full implementation, but
generalized testing with test-cmd and integration has flushed out most
issues

**Comment by timothysc** [comment]
xrefs: https://github.com/kubernetes/kubernetes/issues/22941 
https://github.com/kubernetes/kubernetes/pull/22982

**Comment by wojtek-t** [comment]
Ref #23278

**Comment by bgrant0607** [comment]
FWIW: http://cbor.io/

**Comment by smarterclayton** [comment]
I really like Ion - I hope it takes off.

On Apr 28, 2016, at 7:40 PM, Brian Grant <notifications@github.com> wrote:

FWIW: http://cbor.io/

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/8132#issuecomment-215593835>

**Comment by bgrant0607** [comment]
@smarterclayton @wojtek-t Please ref this issue on related PRs. 

What is the current status?

cc @jbeda

**Comment by smarterclayton** [comment]
Stable serialization reqts.:

* Quantity should be serialized as a string: Provide an int64 version of
Quantity that is much faster
<https://github.com/kubernetes/kubernetes/pull/25243>
* Allow discrimination between unset array and empty array: protobuf
codegen does not handle *[]Foo and *map[Foo]Bar
<https://github.com/kubernetes/kubernetes/issues/25835>

Which block finalizing an output proto IDL we can lock to for 1.3 and
support in a backcompat manner.

**Comment by wojtek-t** [comment]
We finally made a decision to:
1. switch on protobufs by default in 1.3 for communication between components in k8s
[This is done in #25738]
2. decided NOT to switch on protobufs as a storage format in 1.3 release - we will do it in 1.4

**Comment by lavalamp** [comment]
Is there anything left for 1.3 here? I think this is basically done. Let's close this and open a new issue about switching storage to proto by default?

**Comment by smarterclayton** [comment]
I think there are a few things we need to do.

1. Document what our API guarantees are for protobuf
2. Document schema change process
3. Have a follow up issue for the other loose ends (serializing nil vs
empty lists)

**Comment by davidopp** [comment]
Can someone update what work is remaining? This is listed as P1/1.3 which means release blocker.

**Comment by smarterclayton** [comment]
The last items are the remainder.

**Comment by smarterclayton** [comment]
https://github.com/kubernetes/kubernetes.github.io/pull/668 covers doc.
https://github.com/kubernetes/kubernetes/pull/27258 covers optional fields, which has been deferred out of 1.3.  This can be closed.
